### PR TITLE
make ClassJsonAdapter not final

### PR DESCRIPTION
**Pull request by lfarkas**
in this PR i made following changes:

- remove final from ClassJsonAdapter to be able to subclass it
- set includeField as static and moe out from the factory same as isPlatformType since they have totally similar purpose
- move out from the factory and make static the createFieldBindings
- factor out getRawType from the JsonAdapter.Factory.create same as all the above includeField, isPlatformType and createFieldBindings

the reason for these cahnges to be able to subclass from ClassJsonAdapter and easily implement other adapter classes for which i'll send another PR.

### CODE REVIEW

File: moshi/src/main/java/com/squareup/moshi/JsonAdapter.java [hunk 1]
```diff
@@ -37,7 +37,9 @@
   }
 
   public final @Nullable T fromJson(String string) throws IOException {
-    return fromJson(new Buffer().writeUtf8(string));
+    try (Buffer buffer = new Buffer()) {
+    	return fromJson(buffer.writeUtf8(string));
+    }
```
Comment (NightlyNexus) [inline moshi/src/main/java/com/squareup/moshi/JsonAdapter.java]: writing to a Buffer does not IO, and the changes outside of ClassJsonAdapter are unnecessary and unrelated to the change you are trying to make. will you revert these changes please?

Comment (lfarkas) [reply moshi/src/main/java/com/squareup/moshi/JsonAdapter.java]: this way an accident with other code cleanup i like to submit later..

**Comment by lfarkas** [comment]
i was on vacation and i assume i can do it during my holiday so I'll create a new PR only for this and home not to mix different codes.

File: moshi/src/main/java/com/squareup/moshi/JsonAdapter.java (hunk 1 above)
Comment (lfarkas) [reply moshi/src/main/java/com/squareup/moshi/JsonAdapter.java]: anyway Buffer is anAutoClosabe and these should have to be close otherwise code checkers are argue about it

**Comment by swankjesse** [comment]
I don't think we want this. This leaks implementation details and prevents potential future optimizations.

If you want to subclass this, please just copy and paste it into your own project. It isn't that big!

**Comment by lfarkas** [comment]
i write another PR you may close that #364 :-(
but i still not agree with you since eg: isPlatformType should have to be used in all other places otherwise the code becomes inconsistent. the same apply to includeField and getRawType. may be the createFieldBindings can be copied but imho refactoring this function into a separate function makes the code more cleaner.
